### PR TITLE
Avoid unnecessary copies in snippets for file upload

### PR DESCRIPTION
**Pull request by pbugnion**
Prior to this commit, we recommended extracting a string representation of the uploaded file with:

```python
uploaded_file.tobytes().decode("utf-8")
```

This causes a copy of the memoryview to be created when it's converted to bytes. This copy is immediately discarded when it is decoded. Switching to:

```python
codecs.decode(uploaded_file, encoding="utf-8")
```

means that we apparently don't create the intermediate copy (according to Jason in #2721 ). I have not explicitly verified this, though the function signatures at least seem to indicate this.

This commit also adds an example of using the uploaded file in an image widget.

With this PR, the file upload documentation looks like:

Screenshot 2020-01-11 at 12 17 19

Screenshot 2020-01-11 at 12 17 43

### CODE REVIEW

**Comment by jasongrout** [comment]
After #2723, we won't need the tobytes in the image part of the example either.

File: docs/source/examples/Widget List.ipynb [hunk 1]
```diff
@@ -1053,7 +1054,15 @@
     "To convert the uploaded file into a Pandas dataframe, you need to convert it to a StringIO object:\n",
     "\n",
     "```python\n",
-    "pd.read_csv(io.StringIO(uploaded_file[\"content\"].tobytes().decode(\"utf-8\")))\n",
+    "import io\n",
+    "import pandas as pd\n",
+    "pd.read_csv(io.StringIO(codecs.decode(uploaded_file[\"content\"], encoding=\"utf-8\")))\n",
```
Comment (jasongrout) [inline docs/source/examples/Widget List.ipynb]: You can use BytesIO:

```python
pd.read_csv(io.BytesIO(uploaded_file["content"]))
```

File: docs/source/examples/Widget List.ipynb [hunk 2]
```diff
@@ -1036,10 +1036,11 @@
     "#=> b'This is the content of example.txt.\\n'\n",
     "```\n",
     "\n",
-    "To get the contents as a string:\n",
+    "You can get the contents as a string if the file is a text file (more precisely, if the file can be decoded as UTF-8):\n",
```
Comment (jasongrout) [inline docs/source/examples/Widget List.ipynb]:
```suggestion
    "If the file is a text file, you can get the contents as a string by decoding it:\n",
```

**Comment by pbugnion** [comment]
Great catch on using a `BytesIO` object to load the content into a pandas dataframe.

I've made the changes, so I think this is now ready for another review.

File: docs/source/examples/Widget List.ipynb [hunk 3]
```diff
@@ -1053,10 +1054,18 @@
     "    fp.write(uploaded_file[\"content\"])\n",
     "```\n",
     "\n",
-    "To convert the uploaded file into a Pandas dataframe, you need to convert it to a StringIO object:\n",
+    "To convert the uploaded file into a Pandas dataframe, you need to convert it to a [BytesIO object](https://docs.python.org/3/library/io.html#binary-i-o):\n",
```
Comment (jasongrout) [inline docs/source/examples/Widget List.ipynb]:
```suggestion
    "To convert the uploaded file into a Pandas dataframe, you can use a [BytesIO object](https://docs.python.org/3/library/io.html#binary-i-o):\n",
```

**Comment by pbugnion** [comment]
> Thanks, looking good! I added two more very small tweaks, then I think it is good to go.

Thanks for reviewing!

**Pull request merged**